Optitrack pose extractor

naveenoid opened the pull request:
Extracts pose of desired object from an OPTITRACK_FRAME_T message and transforms to the desired reference frame.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6663)
<!-- Reviewable:end -->

naveenoid commented:
+@sammy-tri for feature review. This PR is slightly different from the one I described in our 1-1. The angular velocity estimate will be computed in a different system to be submitted in a follow up PR.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-KpvSLtwZInW0FKkV8c_:b-18bhdd)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+(status: curate commits before merging) first pass complete

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/manipulation/perception/BUILD, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvZbMWb2FEMqXb1sZu:-KpvZbMWb2FEMqXb1sZv:bu12pwa) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/BUILD#L15)):*
> ```Python
>     srcs = ["optitrack_pose_extractor.cc"],
>     hdrs = ["optitrack_pose_extractor.h"],
>     visibility = ["//drake/examples/kuka_iiwa_arm:__subpackages__"],
> ```

I don't think it makes sense to reference visibility in a specific example from a manipulation library.  And wouldn't the desired visibility here be public?

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvahJuv8Ix3C9E-Ko-:-KpvahJvUqLKeitTYOR6:b-x03wcg) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L11)):*
> ```cpp
> #include <iostream>
> #define PRINT_VAR(x) std::cout << #x ": " << x << std::endl;
> #define PRINT_VARn(x) std::cout << #x ":\n" << x << std::endl;
> ```

leftover debugging code?

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 24 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvaq-WGPyJjdP94YBj:-Kpvaq-XvNVrwUY0rsuE:b-k7ul63) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L24)):*
> ```cpp
>     unsigned int object_id, const Isometry3<double>& world_X_optitrack,
>     double optitrack_lcm_status_period) :
> object_id_(object_id), measured_pose_output_port_{
> ```

strange indentation (which remains various types of strange through the rest of the initializers)

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvb-6_zVskrjtdMvJd:-Kpvb-6_zVskrjtdMvJe:bkzgomx) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L30)):*
> ```cpp
>     this->set_name("Optitrack pose extractor");
>     this->DeclareAbstractInputPort();
>     this->DeclareDiscreteState(13);
> ```

BTW 13 doubles is a little long to keep track of with no documentation.  Consider using `lcm_vector_gen.sh` or at least commenting the structure of the data.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 49 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvchGXBk5Eq-j7lTC2:-KpvchGYOEbi8qG5YaCV:b-j5tydf) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L49)):*
> ```cpp
> 
>   DRAKE_THROW_UNLESS(object_id_ < rigid_bodies.size());
>     // The optitrack quaternion ordering is Z-W-X-Y.
> ```

odd indentation change

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvd8BW3FUfDQQurYOQ:-Kpvd8BXEug5q9LyiRiZ:b-ramemi) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L51)):*
> ```cpp
>     // The optitrack quaternion ordering is Z-W-X-Y.
>     Eigen::Quaterniond quaternion(
>             rigid_bodies[object_id_].quat[3],
> ```

Is the space of rigid body IDs always 0..rigid_bodies.size()?  If we're going to make this assumption I think at a minimum we should verify that `rigid_bodies[object_id_].id == object_id_`.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 74 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvdhAgfa15_vmc6ZuC:-KpvdhAhIizD76YRbrAj:b-1ca0bm) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L74)):*
> ```cpp
>   Eigen::VectorBlock<VectorX<double>> measured_pose_output =
>       output->get_mutable_value();
>   measured_pose_output = state_value;
> ```

Didn't we declare a state of size 13 and this output port to be of size 7?

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpva83gD_NbkuTwv_wo:-Kpva83gD_NbkuTwv_wp:bpgzoqx) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.h#L13)):*
> ```objc
> 
> /**
>  * Extracts and provides an output of the pose of a desired object (orientation
> ```

I think instead of saying "orientation in quaternions" it would be better to describe the structure of the vector valued output port (also that it's vector valued).

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvaSED9_e3G8haQeHH:-KpvaSED9_e3G8haQeHI:baz754j) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.h#L33)):*
> ```objc
>   OptitrackPoseExtractor(unsigned int object_id,
>                          const Isometry3<double>& world_X_optitrack,
>   double optitrack_lcm_status_period);
> ```

odd indentation

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvaYYpoIcns87pqbH0:-KpvaYYqv1aPYm2phIA0:baz754j) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.h#L42)):*
> ```objc
>  protected:
>   void OutputMeasuredPose(const systems::Context<double>& context,
>                               systems::BasicVector<double>* output) const;
> ```

odd indentation

---

*[drake/manipulation/perception/test/optitrack_pose_extractor_test.cc, line 4 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvebl0Ipj41PDGZkTs:-Kpvebl1pc-MIWjVnctu:b-8lmf4e) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/test/optitrack_pose_extractor_test.cc#L4)):*
> ```cpp
> 
> #include <gtest/gtest.h>
> #include <memory>
> ```

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes specifies C++ system files before other libraries .h files

---

*[drake/manipulation/perception/test/optitrack_pose_extractor_test.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvevMYUC8N6X2Foebu:-KpvevMYUC8N6X2Foebv:b-dm5so8) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/test/optitrack_pose_extractor_test.cc#L23)):*
> ```cpp
>   void Initialize(int object_id = 0) {
>     dut_ = std::make_unique<OptitrackPoseExtractor>(
>         object_id /* object_id */, Isometry3<double>::Identity()
> ```

I'm concerned that using the identity as the optitrack transform instead of a realistic transform here fails to test a significant function of the code.

---

*[drake/manipulation/perception/test/optitrack_pose_extractor_test.cc, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvfVNB27zyv0Ak3d8N:-KpvfVNCRVd5kqIQvXaG:b-auwnsb) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/test/optitrack_pose_extractor_test.cc#L79)):*
> ```cpp
>   // An arbitrarily chosen test pose is assigned to the default_body.
>   Isometry3<double> test_pose;
>   test_pose.translation()<<0.01, -5.0, 10.10;
> ```

missing whitespace around `<<` operator

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-KpvZX0z2WPLfYPneRR7:bh4v76d)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Added additional unit tests / fixed the non running cpplint and clang-formatted everything.

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/manipulation/perception/BUILD, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvZbMWb2FEMqXb1sZu:-Kpz_h-HvUZKSunAzi23:b-yj1ezx) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/BUILD#L15)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I don't think it makes sense to reference visibility in a specific example from a manipulation library.  And wouldn't the desired visibility here be public?
</blockquote></details>

Done. a git move artifact.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvahJuv8Ix3C9E-Ko-:-Kpz_oOlD3flsLqsQ7ML:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L11)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

leftover debugging code?
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 24 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvaq-WGPyJjdP94YBj:-Kpz_p8LRrI2SwJKEQaI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L24)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

strange indentation (which remains various types of strange through the rest of the initializers)
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvb-6_zVskrjtdMvJd:-KpzcoBn068oRZICmH8b:bzga3qn) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L30)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW 13 doubles is a little long to keep track of with no documentation.  Consider using `lcm_vector_gen.sh` or at least commenting the structure of the data.
</blockquote></details>

Done. 13 was a legacy version error - current version is 7.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 49 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvchGXBk5Eq-j7lTC2:-Kpzct3d6hysiDUQVoim:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L49)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

odd indentation change
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvd8BW3FUfDQQurYOQ:-KpzdB-vDbon079fmegS:b1vsa6b) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L51)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Is the space of rigid body IDs always 0..rigid_bodies.size()?  If we're going to make this assumption I think at a minimum we should verify that `rigid_bodies[object_id_].id == object_id_`.
</blockquote></details>

Done. Added an additional unit test to verify this expected behavior.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 74 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvdhAgfa15_vmc6ZuC:-KpzdCG6LHHCN6xR6GFD:b-5rfa3n) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.cc#L74)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Didn't we declare a state of size 13 and this output port to be of size 7?
</blockquote></details>

Done. fixed now.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpva83gD_NbkuTwv_wo:-Kq-f1q9-P8jAbQSw7vy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.h#L13)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I think instead of saying "orientation in quaternions" it would be better to describe the structure of the vector valued output port (also that it's vector valued).
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvaSED9_e3G8haQeHH:-Kpzdn_dKKZMjSP51tjf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.h#L33)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

odd indentation
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvaYYpoIcns87pqbH0:-Kpzdphz7iiJBaa49keu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/optitrack_pose_extractor.h#L42)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

odd indentation
</blockquote></details>

Done.

---

*[drake/manipulation/perception/test/optitrack_pose_extractor_test.cc, line 4 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kpvebl0Ipj41PDGZkTs:-Kpze3atB-9Bhvdnn2Lq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/test/optitrack_pose_extractor_test.cc#L4)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes specifies C++ system files before other libraries .h files
</blockquote></details>

Done.

---

*[drake/manipulation/perception/test/optitrack_pose_extractor_test.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvevMYUC8N6X2Foebu:-Kpze5ixqH2oqtFBLx_a:b-lxuloy) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/test/optitrack_pose_extractor_test.cc#L23)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'm concerned that using the identity as the optitrack transform instead of a realistic transform here fails to test a significant function of the code.
</blockquote></details>

Done. Added a new test exclusively for this.

---

*[drake/manipulation/perception/test/optitrack_pose_extractor_test.cc, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-KpvfVNB27zyv0Ak3d8N:-Kq-f-Pj81EX2qHh_LqU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/600af1a0fc8ef528d939cc15643315d756181ffb/drake/manipulation/perception/test/optitrack_pose_extractor_test.cc#L79)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

missing whitespace around `<<` operator
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-Kq-fsxQMOfXH8EE0grO:bkxud1t)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +@sherm1 for platform review per rotation

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-Kq8fnSBRuhSSjxK6XLh:b-lkbyk7)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review pass done -- a few comments, PTAL.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq92cBz9abJz1-YKBGo:-Kq92cBz9abJz1-YKBGp:b-4lus32) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.cc#L52)):*
> ```cpp
> 
>   DRAKE_THROW_UNLESS(object_id_ < rigid_bodies.size());
>   DRAKE_THROW_UNLESS(rigid_bodies[object_id_].id == (int)object_id_);
> ```

I'm confused as to why object_id_ is unsigned but here you are casting it to an int. Should it have been an int in the first place?

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 54 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq93YgEykKvdQdBGm80:-Kq93YgEykKvdQdBGm81:b-ecu8zu) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.cc#L54)):*
> ```cpp
>   DRAKE_THROW_UNLESS(rigid_bodies[object_id_].id == (int)object_id_);
> 
>   // The optitrack quaternion ordering is Z-W-X-Y.
> ```

This comment does not match the code. "W" is the scalar component of a quaternion. Eigen's quaternion constructor wants WXYZ order. The code below suggests that the Optitrack ordering is XYZW, not ZWXY as written.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 60 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq931CGb8-xfGZqx55A:-Kq931CGb8-xfGZqx55B:bv9z72n) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.cc#L60)):*
> ```cpp
> 
>   // Transform pose to world frame.
>   Isometry3<double> X_OpOb;
> ```

Please define the "Ob" frame. Is that the rigid body frame? Conventionally that would be denoted just "B", in which case you could write the local transform as X_OB which would be a little tidier. Also you defined the "O" frame in the header but not "Op".

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq94W8upPaD8jFIHiJl:-Kq94W8v7tQ5GyKaKnVt:bxybf4f) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.cc#L66)):*
> ```cpp
>                                          rigid_bodies[object_id_].xyz[2]);
>   X_OpOb.makeAffine();
>   Isometry3<double> X_WOb = X_WOp_ * X_OpOb;
> ```

BTW Consider: `X_WB = X_WO * X_OB`.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq8vTLWOJhvh8ojjt2X:-Kq8vTLXu3c3S5L60Kl4:bwycc9r) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.h#L24)):*
> ```objc
>   /**
>    * Constructs the OptitrackPoseExtractor.
>    * @param object_id : An ID of the object being tracked. This ID must
> ```

BTW don't put ":" after the parameter names here (and below) -- doxygen formats these properly without the colons.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq8waliiBCbtXDCxZ7N:-Kq8waliiBCbtXDCxZ7O:b-6wld1e) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.h#L28)):*
> ```objc
>    * else a runtime exception is thrown.
>    * @param world_X_optitrack : The transformation from world frame (Drake
>    * frame) to the optitrack frame.
> ```

Please see [here](http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html) for description of our unambiguous "monogram" notation for transforms. I believe (but can't be sure) that what you are asking for here is `X_WO` but the wording in the description (though ambiguous at best) would conventionally describe `X_OW` which is the inverse transform.

Please check to see exactly how this transform should be defined and use the standard notation so that there cannot be any misinterpretation. Assuming you want `X_WO`, here are a few ways to document it, depending on taste:
```c++
/**
 * @param X_WO  
 *     The pose of the Optitrack frame O in the World frame W.
 * @param optitrack_frame_in_world 
 *     The pose X_WO of the Optitrack frame O in the World frame W.
 */
```
Although naming the parameter X_WO is cryptic, it is completely unambiguous and clearest to those familiar with our notation, so I prefer that. The other one is  newbie-friendlier perhaps, but requires reading the comment to know exactly what it means. In any case there is no good unambiguous way to describe this in English so we would want to refer said newbie to our notation page anyway.

(Edit: later I saw that the data member is clearly X_WO so you obviously know the right notation! Just need to make the call here unambiguous too.)

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 33 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq8vuGNSCW92xXedmaJ:-Kq8vuGNSCW92xXedmaK:bhu3fi1) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.h#L33)):*
> ```objc
>    * optitrack messages.
>    */
>   OptitrackPoseExtractor(unsigned int object_id,
> ```

BTW styleguide doesn't permit unsigned integers except when forced for compatibility reasons. Is that the case here? I don't suppose there is a typedef for this defined with the Optitrack message you could use? If not, consider making one here (like `OptitrackIdType` or whatever) so it is clear where the type requirement comes from. 

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9-klx7xPumfMqA0so:-Kq9-klx7xPumfMqA0sp:ba1f7wk) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.h#L42)):*
> ```objc
> 
>  protected:
>   void OutputMeasuredPose(const systems::Context<double>& context,
> ```

BTW this should be private. Also, took me a while to figure out this was the output port Calc() method -- a short comment to that effect would be helpful.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 54 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq90TI0JWbfwviM34Sl:-Kq90TI0JWbfwviM34Sm:b-bkf821) ([raw file](https://github.com/robotlocomotion/drake/blob/1c0b5f7e75381d3ceb2342764cbd5d65f7997e11/drake/manipulation/perception/optitrack_pose_extractor.h#L54)):*
> ```objc
>   const int measured_pose_output_port_{-1};
>   // Pose of the optitrack frame O in the world frame W.
>   const Isometry3<double> X_WOp_;
> ```

BTW oh, this is good here! Should be unambiguous in the constructor also.

BTW why the "Op" here? The frame name is apparently just "O" so the symbol should be X_WO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-Kq94thkIRNH_9Hwwjms:b-octksh)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Thanks! Addressed the comments / defects. 

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq92cBz9abJz1-YKBGo:-Kq9gJ7QYwQpKPaOswxh:b-7tzaa1) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.cc#L52)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I'm confused as to why object_id_ is unsigned but here you are casting it to an int. Should it have been an int in the first place?
</blockquote></details>

Done. its now an int.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 54 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq93YgEykKvdQdBGm80:-Kq9gXhdIi8VxywzPcR2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.cc#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This comment does not match the code. "W" is the scalar component of a quaternion. Eigen's quaternion constructor wants WXYZ order. The code below suggests that the Optitrack ordering is XYZW, not ZWXY as written.
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 60 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq931CGb8-xfGZqx55A:-Kq9gv5krpGmXZreyzB8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.cc#L60)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Please define the "Ob" frame. Is that the rigid body frame? Conventionally that would be denoted just "B", in which case you could write the local transform as X_OB which would be a little tidier. Also you defined the "O" frame in the header but not "Op".
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq94W8upPaD8jFIHiJl:-Kq9gveRXhcH3gmvFBd5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.cc#L66)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW Consider: `X_WB = X_WO * X_OB`.
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq8vTLWOJhvh8ojjt2X:-Kq9h-vVVk_3aHfsUb4u:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.h#L24)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW don't put ":" after the parameter names here (and below) -- doxygen formats these properly without the colons.
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq8waliiBCbtXDCxZ7N:-Kq9qC4nmsrnQS7vs_Rf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Please see [here](http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html) for description of our unambiguous "monogram" notation for transforms. I believe (but can't be sure) that what you are asking for here is `X_WO` but the wording in the description (though ambiguous at best) would conventionally describe `X_OW` which is the inverse transform.

Please check to see exactly how this transform should be defined and use the standard notation so that there cannot be any misinterpretation. Assuming you want `X_WO`, here are a few ways to document it, depending on taste:
```c++
/**
 * @param X_WO  
 *     The pose of the Optitrack frame O in the World frame W.
 * @param optitrack_frame_in_world 
 *     The pose X_WO of the Optitrack frame O in the World frame W.
 */
```
Although naming the parameter X_WO is cryptic, it is completely unambiguous and clearest to those familiar with our notation, so I prefer that. The other one is  newbie-friendlier perhaps, but requires reading the comment to know exactly what it means. In any case there is no good unambiguous way to describe this in English so we would want to refer said newbie to our notation page anyway.

(Edit: later I saw that the data member is clearly X_WO so you obviously know the right notation! Just need to make the call here unambiguous too.)
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 33 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq8vuGNSCW92xXedmaJ:-Kq9qHVd0xEYFrEqgfZC:buz32u9) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.h#L33)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW styleguide doesn't permit unsigned integers except when forced for compatibility reasons. Is that the case here? I don't suppose there is a typedef for this defined with the Optitrack message you could use? If not, consider making one here (like `OptitrackIdType` or whatever) so it is clear where the type requirement comes from. 
</blockquote></details>

Done. Changed to int. 

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9-klx7xPumfMqA0so:-Kq9rdm4wUht_j4xjn5s:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.h#L42)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this should be private. Also, took me a while to figure out this was the output port Calc() method -- a short comment to that effect would be helpful.
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 54 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq90TI0JWbfwviM34Sl:-Kq9rpLr420z3siCrcZX:bs7wumx) ([raw file](https://github.com/robotlocomotion/drake/blob/fc1c2492b0f0f3e3c273d1af42ccc82ca1abb0f0/drake/manipulation/perception/optitrack_pose_extractor.h#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW oh, this is good here! Should be unambiguous in the constructor also.

BTW why the "Op" here? The frame name is apparently just "O" so the symbol should be X_WO.
</blockquote></details>

Done. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-Kq9s7FfbNtbomX9wTV2:b-vz8kv4)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks, Naveen! Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending a few minor BTWs below.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9zyyCartEhQKh-Xfc:-Kq9zyyCartEhQKh-Xfd:b5rwyvi) ([raw file](https://github.com/robotlocomotion/drake/blob/5f954f0fb1384302874474b84ab6d0003cbd62b3/drake/manipulation/perception/optitrack_pose_extractor.cc#L27)):*
> ```cpp
>                   &OptitrackPoseExtractor::OutputMeasuredPose)
>               .get_index()},
>       X_WO(X_WO) {
> ```

BTW Yikes -- scary! Should be `X_WO_(X_WO)` so the data member and argument have different names.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9zh0y4cZsHJEIM0yj:-Kq9zh0y4cZsHJEIM0yk:b-4hmzkl) ([raw file](https://github.com/robotlocomotion/drake/blob/5f954f0fb1384302874474b84ab6d0003cbd62b3/drake/manipulation/perception/optitrack_pose_extractor.cc#L67)):*
> ```cpp
>                                          rigid_bodies[object_id_].xyz[2]);
>   X_OB.makeAffine();
>   Isometry3<double> X_WOb = X_WO * X_OB;
> ```

BTW I think you wanted `X_WB =` here?

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9zUIBd75lziHak7Tz:-Kq9zUIBd75lziHak7U-:bvtm3sv) ([raw file](https://github.com/robotlocomotion/drake/blob/5f954f0fb1384302874474b84ab6d0003cbd62b3/drake/manipulation/perception/optitrack_pose_extractor.h#L53)):*
> ```objc
>   const int measured_pose_output_port_{-1};
>   // Pose of the optitrack frame O in the world frame W.
>   const Isometry3<double> X_WO;
> ```

BTW data member is missing its trailing "_" now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-KqA-BxWx9VF6UBAFtkg:b-gzbuhr)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9zyyCartEhQKh-Xfc:-KqA4HcoTaVSO7mps6ed:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5f954f0fb1384302874474b84ab6d0003cbd62b3/drake/manipulation/perception/optitrack_pose_extractor.cc#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW Yikes -- scary! Should be `X_WO_(X_WO)` so the data member and argument have different names.
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.cc, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9zh0y4cZsHJEIM0yj:-KqA4GnlxwVrA6kFnJ1D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5f954f0fb1384302874474b84ab6d0003cbd62b3/drake/manipulation/perception/optitrack_pose_extractor.cc#L67)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I think you wanted `X_WB =` here?
</blockquote></details>

Done.

---

*[drake/manipulation/perception/optitrack_pose_extractor.h, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-Kq9zUIBd75lziHak7Tz:-KqA4Fm3adrdgeLeeWf9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5f954f0fb1384302874474b84ab6d0003cbd62b3/drake/manipulation/perception/optitrack_pose_extractor.h#L53)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW data member is missing its trailing "_" now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
moving to squash merge. 
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6663#-:-KqAHTBWiKkh6D5WCDa9:bn9w675)*
<!-- Sent from Reviewable.io -->
